lib/fog/libvirt/requests/compute/list_domains.rb

```diff
@@ -40,10 +40,15 @@ def domain_display xml
 
         def domain_volumes xml
           vols_by_file = xml_elements(xml, "domain/devices/disk/source", "file")
+          vols_by_dev = xml_elements(xml, "domain/devices/disk/source", "dev")
           vols_by_name = xml_elements(xml, "domain/devices/disk/source", "name")
           vols = []
-          vols_by_file.zip(vols_by_name).each do |by_file,by_name|
-            vols.push(by_file.nil? ? by_name : by_file)
+          vols_by_file.zip(vols_by_dev,vols_by_name).each do |by_file,by_dev,by_name|
+            if by_file.nil?
+              vols.push(by_dev.nil? ? by_name : by_dev)
+            else
+              vols.push(by_file)
+            end
```

Comment:
Just thinking out loud. Isn't this equivalent?
```suggestion
            vols.push(by_file || by_dev || by_name)
```

And you could even go a step further and replace `.each` with `.map`, then get rid of the `vols` variable altogether, but that may be too much of a refactor for this patch.